test/suite/functional/Meta/ComposerJson/composer_default.json

```diff
@@ -10,13 +10,10 @@
     "require": {
         "php": ">=7.2",
         "docler-labs/api-client-exception": "^1.0",
-        "guzzlehttp/guzzle": "^7.1",
         "guzzlehttp/psr7": "^1.6",
         "pimple/pimple": "^3.3",
         "psr/container": "^1.0",
-        "psr/http-client": "^1.0",
-        "psr/http-client-implementation": "^1.0",
```

Comment:
Please double check, this psr/http-client-implementation virtual package is useful in my opinion. It explicitly states that the client would not work without actual implementation of psr/http-client

Reply:
we can't have it there :( I just tried it but I get an error:
```
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - The requested package psr/http-client-implementation could not be found in any version, there may be a typo in the package name.
```
if you add a virtual package, composer will force you to add a correspondent implementation package too, and idea with decoupling client from guzzle is to have client working with PSR interfaces only and add guzzle dependency where it's using the client instead...

Reply:
I think you misunderstood the purpose of the virtual package. It forces to use any psr/http-client implementation upon the client usage, so when you require the package, not in composer.json of the client itself.
It makes sense if you think about it: you can not use the client library without psr/http-client implementation, it is mandatory. You decouple it, that is nice, this virtual package does not prevent it, it supplements that idea. Further on it would be possible to discover the HTTP-client implementation automatically: https://github.com/php-http/documentation/blob/master/discovery.rst
Hovewer, it is just a suggestion, you can skip it.

Reply:
hmm I'll double check it, maybe I misunderstood it indeed, the idea seems nice... and the discovery thing, I like it very much, I even thought about that as a follow up too couple days ago 😄 but couldn't figure it out still a way to configure the client properly through the discovery mechanism (like timeout, base uri, stack etc.)

Reply:
same situation:
```
$ composer install
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - The requested package psr/http-client-implementation could not be found in any version, there may be a typo in the package name.

Potential causes:
 - A typo in the package name
 - The package is not available in a stable-enough version according to your minimum-stability setting
   see <https://getcomposer.org/doc/04-schema.md#minimum-stability> for more details.
 - It's a private package and you forgot to add a custom repository to find it
```
If you add `psr/http-client-implementation` to `composer.json` of the client, it expects an implementation to be there as well, otherwise we can't `composer install/update/etc.` on the client. Who's using the client will add the implementation and there composer will probably work, but adding this constraint on the client itself will "break" composer there...

Reply:
Could you explain how are you checking?
You need to have an app with a generated api client and for example, Guzzle as dependencies. Then it would work, and that is a real-life use case.
If you need to do composer install to run tests, you may add Guzzle(or any other base client) as a dev-dependency.  But why writing tests for generated code?
Am I missing something?

Reply:
ok, I've googled it, and seems people who uses virtual package indeed have the same "issue" 😄 
```
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - The requested package phpcr/phpcr-implementation could not be found in any version, there may be a typo in the package name.

Potential causes:
 - A typo in the package name
 - The package is not available in a stable-enough version according to your minimum-stability setting
   see <https://getcomposer.org/doc/04-schema.md#minimum-stability> for more details.
 - It's a private package and you forgot to add a custom repository to find it

Read <https://getcomposer.org/doc/articles/troubleshooting.md> for further common problems.
```
(if you try `composer install` on https://github.com/doctrine/phpcr-odm, for example).

It's as you said, it works fine on the app which is requiring the client... but composer on the client itself won't work, but it's ok I guess. As you said, we ain't writing tests there, and at least this way we can "force" who's using the client to have http-client-implementation package before requiring the autogenerated client itself.

There's also "suggests", which can be used to hint instead of blocking... then at least composer wouldn't fail on the client, but I'm ok with either, wdyt?

(meanwhile I'm pushing commit with require http-client-implementation in case we go this route)

Reply:
I understand some may find it annoying, but either way, it would not work if you don't have guzzle(or similar) in requirements. In my opinion, it is more obvious that way, however, I understand that it is controversial.